spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeIntoExec.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.expressions.BasePredicate
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.expressions.UnsafeProjection
+import org.apache.spark.sql.catalyst.expressions.codegen.GeneratePredicate
+import org.apache.spark.sql.catalyst.plans.logical.MergeIntoParams
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.UnaryExecNode
+
+case class MergeIntoExec(
+    mergeIntoParams: MergeIntoParams,
+    @transient targetRelation: DataSourceV2Relation,
+    override val child: SparkPlan) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] = targetRelation.output
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    child.execute().mapPartitions {
+      processPartition(mergeIntoParams, _)
+    }
+  }
+
+  private def generateProjection(exprs: Seq[Expression], attrs: Seq[Attribute]): UnsafeProjection = {
+    UnsafeProjection.create(exprs, attrs)
+  }
+
+  private def generatePredicate(expr: Expression, attrs: Seq[Attribute]): BasePredicate = {
+    GeneratePredicate.generate(expr, attrs)
+  }
+
+  def applyProjection(
+     actions: Seq[(BasePredicate, UnsafeProjection)],
+     projectTargetCols: UnsafeProjection,
+     projectDeleteRow: UnsafeProjection,
+     inputRow: InternalRow,
+     targetRowNotPresent: Boolean): InternalRow = {
+
+    /**
+     * Find the first combination where the predicate evaluates to true.
+     * In case when there are overlapping condition in the MATCHED
+     * clauses, for the first one that satisfies the predicate, the
+     * corresponding action is applied. For example:
+     *
+     * WHEN MATCHED AND id > 1 AND id < 10 UPDATE *
+     * WHEN MATCHED AND id = 5 OR id = 21 DELETE
+     *
+     * In above case, when id = 5, it applies both that matched predicates. In this
+     * case the first one we see is applied.
+     */
```

Comment:
Style: this is a doc comment because it starts with `/**`. Usually, multi-line comments in code would use `//  ` on each line.